quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,885 @@
+        // Extract the local storage root path, also validate that local storage
+        // is of type [`LocalFileStorage]
+        let local_storage_root_path =
+            LocalFileStorage::extract_root_path_from_uri(&local_storage.uri())?;
+
+        Ok(Self {
+            remote_storage,
+            local_storage,
+            local_storage_root: local_storage_root_path,
+            disk_capacity: DiskCapacity {
+                max_num_files,
+                max_num_bytes,
+            },
+            disk_cache: Mutex::new(WrappedCache {
+                lru: LruCache::unbounded(),
+                num_bytes: 0,
+                num_files: 0,
+            }),
+        })
+    }
+
+    /// Create an instance of [`StorageWithUploadCache`] from a path.
+    ///
+    /// It needs a folder `path` previously used by an instance of [`StorageWithUploadCache`].
+    /// Lastly, the file at `{path}/[`CACHE_STATE_FILE_NAME`]` should be present with valid content.
+    pub fn from_path(
+        remote_storage: Arc<dyn Storage>,
+        storage_uri_resolver: &StorageUriResolver,
+        path: &Path,
+    ) -> StorageResult<Self> {
+        let cache_state = CacheState::from_path(path)?;
+        let local_storage = storage_uri_resolver
+            .resolve(&cache_state.local_storage_uri)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+
+        // Verify that the provided `root_path` and extracted `root_path` match.
+        let local_storage_root_path =
+            LocalFileStorage::extract_root_path_from_uri(&cache_state.local_storage_uri)?;
+        if local_storage_root_path != path.to_path_buf() {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Invalid root path: {}, but {} expected ",
+                path.to_string_lossy(),
+                local_storage_root_path.to_string_lossy()
+            )));
+        }
+
+        let num_files = cache_state.items.len();
+        let num_bytes = cache_state.items.iter().map(|(_, size)| *size).sum();
+        if cache_state
+            .disk_capacity
+            .exceeds_capacity(num_bytes, num_files)
+        {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Initial cache items should not exceed cache capacity.",
+            )));
+        }
+
+        let mut disk_cache = LruCache::unbounded();
+        for (path, size) in cache_state.items {
+            disk_cache.put(path, size);
+        }
+
+        Ok(Self {
+            remote_storage,
+            local_storage,
+            local_storage_root: local_storage_root_path,
+            disk_capacity: cache_state.disk_capacity,
+            disk_cache: Mutex::new(WrappedCache {
+                lru: LruCache::unbounded(),
+                num_bytes,
+                num_files,
+            }),
+        })
+    }
+
+    /// Check if an incomming item will exceed the capacity once inserted.
```

Comment:
Comment is misleading. Here we need to pass the total num bytes and total num files to check if we exceed capacity.
/// Returns false if `num_bytes` or `num_files` are greater than disk capacity.

Reply:
I feel the comment is ok. but probably the params names
```rust
fn exceeds_capacity(&self, total_bytes_after_insert: usize, total_files_after_insert: usize) -> bool {
        self.disk_capacity.exceeds_capacity(total_bytes_after_insert, total_files_after_insert)
    }
```

Reply:
I feel the function is not super helpful :)